proof-systems/src/darlin/tests/final_darlin.rs

```diff
@@ -0,0 +1,353 @@
+            || "c",
+            || self.c.ok_or(SynthesisError::AssignmentMissing)
+        )?;
+        let d = cs.alloc_input(
+            || "d",
+            || self.d.ok_or(SynthesisError::AssignmentMissing)
+        )?;
+
+        for i in 0..(self.num_variables - 5 - (2 * deferred_len)) {
+            let _ = cs.alloc(
+                || format!("var {}", i),
+                || self.a.ok_or(SynthesisError::AssignmentMissing),
+            )?;
+        }
+
+        for i in 0..(self.num_constraints - 1 - test_constraints){
+            cs.enforce(
+                || format!("constraint {}", i),
+                |lc| lc + a,
+                |lc| lc + b,
+                |lc| lc + c,
+            );
+        }
+        cs.enforce(
+            || format!("constraint {}", self.num_constraints - 1),
+            |lc| lc + c,
+            |lc| lc + b,
+            |lc| lc + d,
+        );
+
+        Ok(())
+    }
+}
+
+impl<G1, G2> PCDCircuit<G1> for TestCircuit<G1, G2>
+    where
+        G1: AffineCurve<BaseField = <G2 as AffineCurve>::ScalarField> + ToConstraintField<<G2 as AffineCurve>::ScalarField>,
+        G2: AffineCurve<BaseField = <G1 as AffineCurve>::ScalarField> + ToConstraintField<<G1 as AffineCurve>::ScalarField>,
+{
+    type SetupData       =  CircuitInfo<G1, G2>;
+    type IncrementalData =  (G1::ScalarField, G1::ScalarField);
+    type SystemInputs    =  FinalDarlinDeferredData<G1, G2>;
+    type PreviousPCD     =  TestPrevPCD<G1, G2>;
+
+    fn init(config: Self::SetupData) -> Self {
+        Self {
+            a: None,
+            b: None,
+            c: None,
+            d: None,
+            num_constraints: config.num_constraints,
+            num_variables: config.num_variables,
+            deferred: config.dummy_deferred.clone()
+        }
+    }
+
+    fn init_state(
+        config:               Self::SetupData,
+        previous_proofs_data: Vec<Self::PreviousPCD>,
+        _previous_proofs_vks: Vec<<Self::PreviousPCD as PCD>::PCDVerifierKey>,
+        incremental_data:     Self::IncrementalData
+    ) -> Self
+    {
+        assert_eq!(previous_proofs_data.len(), 1);
+
+        let a = incremental_data.0;
+        let b = incremental_data.1;
+
+        let mut c = a;
+        c.mul_assign(&b);
+        let mut d = c;
+        d.mul_assign(&b);
+
+        Self {
+            a: Some(a),
+            b: Some(b),
+            c: Some(c),
+            d: Some(d),
+            num_constraints: config.num_constraints,
+            num_variables: config.num_variables,
+            deferred: previous_proofs_data[0].0.clone(),
+        }
+    }
+
+    fn get_sys_ins(&self) -> Result<&Self::SystemInputs, PCDError> {
+        Ok(&self.deferred)
+    }
+
+    fn get_usr_ins(&self) -> Result<Vec<G1::ScalarField>, PCDError> {
+        let c = self.c.ok_or(PCDError::MissingUserInputs("c".to_owned()))?;
+        let d = self.d.ok_or(PCDError::MissingUserInputs("d".to_owned()))?;
+        Ok(vec![c, d])
+    }
```

Comment:
The functions ``get_sys_ins()`` and ``get_usr_ins()`` do not run the circuit at all.
Is this representative for the usual case? (How does the synthesizer work if public 
inputs are considered as  outputs of a computation? Do we always need to compute
such inputs outside the circuit before we can run the synthesizer?)

Reply:
Currently the Synthesizer can't return anything. If we want to do it we need to find some workarounds.
This means that, for the moment, we are forced to pass to the synthesizer all the data (including the one that is supposed to be "computed" while running the circuit)

Reply:
Let's put that on our list!